src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/AwsSdk/LambdaInvokeRequestHandler.cs

```diff
@@ -0,0 +1,99 @@
+// Copyright 2020 New Relic, Inc. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+using System.Collections.Concurrent;
+using System;
+using System.Threading.Tasks;
+using NewRelic.Agent.Api;
+using NewRelic.Agent.Api.Experimental;
+using NewRelic.Agent.Extensions.Providers.Wrapper;
+using NewRelic.Reflection;
+using NewRelic.Agent.Extensions.Helpers;
+
+namespace NewRelic.Providers.Wrapper.AwsSdk
+{
+    internal static class LambdaInvokeRequestHandler
+    {
+        private static ConcurrentDictionary<Type, Func<object, object>> _getResultFromGenericTask = new();
+
+        private static object GetTaskResult(object task)
+        {
+            if (((Task)task).IsFaulted)
+            {
+                return null;
+            }
+
+            var getResponse = _getResultFromGenericTask.GetOrAdd(task.GetType(), t => VisibilityBypasser.Instance.GeneratePropertyAccessor<object>(t, "Result"));
+            return getResponse(task);
+        }
+
+        private static void SetRequestIdIfAvailable(IAgent agent, ITransaction transaction, dynamic response)
+        {
+            try
+            {
+                dynamic metadata = response.ResponseMetadata;
+                string requestId = metadata.RequestId;
+                transaction.AddCloudSdkAttribute("aws.requestId", requestId);
+            }
+            catch (Exception e)
+            {
+                agent.Logger.Debug(e, "Unable to get RequestId from response metadata.");
+            }
+        }
+
+        public static AfterWrappedMethodDelegate HandleInvokeRequest(InstrumentedMethodCall instrumentedMethodCall, IAgent agent, ITransaction transaction, dynamic request, bool isAsync, string region)
+        {
+            string functionName = request.FunctionName;
+            string qualifier = request.Qualifier;
+            if (!string.IsNullOrEmpty(qualifier) && !functionName.EndsWith(qualifier))
+            {
+                functionName = $"{functionName}:{qualifier}";
+            }
+            string arn = AwsSdkHelpers.ConstructArn(agent, functionName, region, "");
```

Comment:
Is the arn going to be a constant value? If so, could we cache it?

Reply:
Individual Lambda ARNs are constant, but the user might be calling more than one. We could do a name-to-ARN map, but I can see an edge case where it might fail: if you were calling the same Lambda in more than one region but only specifying the name each time. That's admittedly pretty unlikely, though.

Reply:
It just seems like the parsing logic is so complicated that it may impact performance if we don't cache the result. This will get called on every Lambda invocation, right?

Reply:
Fair point, done.